File: frontend/src/types/metrics.ts
/** Types for the in-app Reports/metrics dashboard — mirrors
 * `backend/app/schemas/metrics.py`'s `MetricsReportResponse`.
 */

export interface MetricsOverview {
  total_workflows: number;
  completed_workflows: number;
  completed_runs: number;
  total_llm_calls: number;
  total_tokens: number;
  total_cost_usd: number;
  avg_latency_ms: number;
  indexed_repositories: number;
  total_graph_nodes: number;
  total_graph_edges: number;
}

export interface CostByDayPoint {
  day: string;
  cost_usd: number;
  tokens: number;
}

export interface ProviderCost {
  provider: string;
  calls: number;
  cost_usd: number;
  tokens: number;
}

export interface StageCost {
  stage: string;
  calls: number;
  cost_usd: number;
  tokens: number;
}

export interface ModelUsage {
  model: string;
  provider: string;
  calls: number;
  cost_usd: number;
}

export interface RunStageOutcome {
  stage: string;
  total: number;
  succeeded: number;
  failed: number;
}

export interface RepositoryComponentCount {
  repository_id: string;
  name: string;
  components: number;
}

export interface WorkflowSummary {
  id: string;
  title: string;
  status: string;
  current_stage: string;
  workflow_type: string;
  created_at: string;
  updated_at: string;
  cost_usd: number;
  tokens: number;
}

export type MetricsScope = "user" | "global";

export interface MetricsReportResponse {
  scope: MetricsScope;
  generated_at: string;
  window_days: number;
  overview: MetricsOverview;
  cost_by_day: CostByDayPoint[];
  cost_by_provider: ProviderCost[];
  cost_by_stage: StageCost[];
  model_usage: ModelUsage[];
  run_success_by_stage: RunStageOutcome[];
  repository_components: RepositoryComponentCount[];
  recent_workflows: WorkflowSummary[];
}

File: backend/app/schemas/metrics.py
"""Response schemas for the in-app Reports/metrics dashboard.

Mirrors the sections produced by `scripts/generate_report.py` (the original
standalone HTML report), so the live `/reports` page can reach full parity
with that script's output without running it.
"""

from pydantic import BaseModel


class MetricsOverview(BaseModel):
    total_workflows: int
    completed_workflows: int
    completed_runs: int
    total_llm_calls: int
    total_tokens: int
    total_cost_usd: float
    avg_latency_ms: float
    indexed_repositories: int
    total_graph_nodes: int
    total_graph_edges: int


class CostByDayPoint(BaseModel):
    day: str
    cost_usd: float
    tokens: int


class ProviderCost(BaseModel):
    provider: str
    calls: int
    cost_usd: float
    tokens: int


class StageCost(BaseModel):
    stage: str
    calls: int
    cost_usd: float
    tokens: int


class ModelUsage(BaseModel):
    model: str
    provider: str
    calls: int
    cost_usd: float


class RunStageOutcome(BaseModel):
    stage: str
    total: int
    succeeded: int
    failed: int


class RepositoryComponentCount(BaseModel):
    repository_id: str
    name: str
    components: int


class WorkflowSummary(BaseModel):
    id: str
    title: str
    status: str
    current_stage: str
    workflow_type: str
    created_at: str
    updated_at: str
    cost_usd: float
    tokens: int


class MetricsReportResponse(BaseModel):
    scope: str  # "user" | "global"
    generated_at: str
    window_days: int
    overview: MetricsOverview
    cost_by_day: list[CostByDayPoint]
    cost_by_provider: list[ProviderCost]
    cost_by_stage: list[StageCost]
    model_usage: list[ModelUsage]
    run_success_by_stage: list[RunStageOutcome]
    repository_components: list[RepositoryComponentCount]
    recent_workflows: list[WorkflowSummary]


class WorkflowStageLLMUsage(BaseModel):
    """One workflow's LLM consumption for a single stage (Planning,
    Development, Testing, ...) - the per-stage breakdown Frontier Agent
    users need to see which stage actually drove cost/latency, not just
    the workflow's total."""

    stage: str
    models: list[str]
    calls: int
    input_tokens: int
    output_tokens: int
    total_tokens: int
    cost_usd: float
    avg_latency_ms: float


class WorkflowLLMUsageResponse(BaseModel):
    workflow_id: str
    workflow_title: str
    stages: list[WorkflowStageLLMUsage]
